packages/devtools_app/lib/src/memory/flutter/memory_chart.dart

```diff
@@ -0,0 +1,332 @@
+          xAxis
+            ..position = XAxisPosition.BOTTOM
+            ..textSize = 10
+            ..textColor = ColorUtils.WHITE
+            ..drawAxisLine = false
+            ..drawGridLines = true
+            ..textColor = const Color.fromARGB(255, 0, 0, 0)
+            ..centerAxisLabels = true
+            ..setGranularity(1)
+            ..setValueFormatter(XAxisFormatter());
+        },
+        legendSettingFunction: (legend, controller) {
+          legend.enabled = false;
+          // TODO(terry): Need to support legend with a smaller text size.
+          /*
+           legend
+            ..shape = LegendForm.LINE
+            ..verticalAlignment = LegendVerticalAlignment.TOP
+            ..enabled = true
+            ..orientation = LegendOrientation.HORIZONTAL
+            ..typeface = legendTypeFace
+            ..xOffset = 20
+            ..drawInside = false
+            ..horizontalAlignment = LegendHorizontalAlignment.CENTER
+            ..textSize = 6.0;
+          */
+        },
+        highLightPerTapEnabled: true,
+        backgroundColor: ColorUtils.WHITE,
+        drawGridBackground: false,
+        dragXEnabled: true,
+        dragYEnabled: true,
+        scaleXEnabled: true,
+        scaleYEnabled: true,
+        pinchZoomEnabled: false,
+        description: desc);
+
+    // Compute padding around chart.
+    _chartController.setViewPortOffsets(50, 10, 10, 30);
+  }
+
+  final List<Entry> _used = <Entry>[];
+  final List<Entry> _capacity = <Entry>[];
+  final List<Entry> _externalHeap = <Entry>[];
+
+  var _img;
+
+  LineDataSet usedHeapSet;
+  LineDataSet capacityHeapSet;
+  LineDataSet externalMemorySet;
+
+  _loadImage() async => ImageLoader.loadImage('assets/img/star.png');
+
+  int timerDataIndex = 0;
+  Timer _timer;
+  Future<void> startTimer(
+    List<Entry> capacity,
+    List<Entry> used,
+    List<Entry> externalHeap,
+  ) async {
+    if (_img == null) await _loadImage();
+
+    used.clear();
+    capacity.clear();
+    externalHeap.clear();
+
+    // Prime the data.
+    if (timerDataIndex < externalMemoryData.length) {
+      int x;
+      int y;
+
+      x = externalMemoryData[timerDataIndex];
+      y = externalMemoryData[timerDataIndex + 1];
+      externalHeap.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+      x = usedHeapData[timerDataIndex];
+      y = usedHeapData[timerDataIndex + 1] +
+          externalMemoryData[timerDataIndex + 1];
+      used.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+      x = heapCapacityData[timerDataIndex];
+      y = heapCapacityData[timerDataIndex + 1];
+      capacity.add(Entry(x: x.toDouble(), y: y.toDouble(), icon: _img));
+
+      timerDataIndex += 2;
+    }
+
+    // Average rate is ~500-600 ms?
+    _timer = Timer.periodic(const Duration(milliseconds: 200), (Timer timer) {
+      // Pause pressed stop pumping out data simulating a live feed.
+      if (widget.memoryController.paused) return;
```

Comment:
This is a case where we'd want to follow a notifier pattern.  We shouldn't have to repeatedly poll the controller for whether it is paused or not, nor should we have to poll it for whether it has received new data.

Instead, we'd want to have a controller expose the following API:

```dart
Controller {
  Notifier pauseChanged;
  bool get paused => _paused;
  set paused(bool nowPaused) {
    _paused = nowPaused;
    pauseNotifier.notifyAll();
  }

  Notifier memoryDataChanged;
  final List memoryData;
  // Similar to above, notify every time memoryData is added to.
}
```

Then the state would be this:

```dart
class State ... {
  void initState() {
    controller.pauseChanged.addListener(_onPaused);
    controller.memoryDataChanged.addListener(_doMathWithNewData);
  }
}
```

Reply:
Right did this also in the latest code.